### ISSUE TITLE
4.5.10 - reduction expressions (all new)

### ISSUE THREAD

**Issue by pat-rogers** [pat-rogers/Ada-202x-WG9-Informal-Review#139]
Looking more closely at value_sequences involves looking at
iterated_element_associations. And that leads to wondering about whether
the accessibility level is well defined for the loop parameter of such an
iteration. Consider this expression:
   [for Loop_Param of ... => Loop_Param.Aliased_Component'Access]'Reduce
      (Function_With_Anonymous_Access_Parameter, Init_Value)

When Function_With_Anonymous_Access_Parameter is called (at runtime), an
accessibility level will need to be passed in. What rules define
that level?

Perhaps the apparently-missing accessibility level definition follows
from some equivalence rule, but I don't see it.

If there is an issue here, then (at least) two solutions come to mind.
One approach is to make sure this case is well-defined in 3.10.2.
Alternatively, we could require that the Value_Type and the Accum_Type
of a reduction expression shall each not be an anonymous access type.

Is there a similar issue with quantified expressions? Perhaps not
(although it seems suspicious that 5.5.2(8.a/5) talks about "the loop
statement", as opposed to other looping constructs).

This is a little off the subject of 4.5.10, but there seems to be
a related accessibility problem with generalized iterators.

The RM is not as clear as I feel it should be that the execution
of a loop statement with a generalized iterator does *not* have a single
loop parameter object that takes on different values; each iteration
gets its own loop parameter object (or at least it should).

Contrast this with RM 5.5.2:
   8.a/5
   The loop parameter of a generalized iterator has the same accessibility
   as the loop statement. This means that the loop parameter object is
   finalized when the loop statement is left. (It also may be finalized
   as part of assigning a new value to the loop parameter.)

That wording, at least taken in isolation, would seem to incorrectly
suggest that the loop parameter is finalized even if there
are no iterations.

More importantly, this suggests that if the value of
Some_Part_Of_The_Loop_Parameter'Unchecked_Access is generated in
one iteration and then dereferenced  in the next iteration, then this
does not trigger the 13.11.2 rule
  Evaluating a name that denotes a nonexistent object ...  is erroneous.

But it should. That is, such an execution should be erroneous.

And similarly for 13.11.2 (15.2/4).

The accessibility level for a loop parameter, at least in this case,
should be that of one loop iteration, not that of the entire loop
statement (i.e., all iterations). OTOH, if we are iterating over the
elements of an array with an array component iterator then we want
the accessibility level of the element to be that of the array.

SteveB

**Comment by ARG-Editor** [comment]
This can't possibly be something that is going to matter to 99.99% of reduction expressions, so resolving this sort of corner case can wait for a future Ada. (Preferably about Ada 2222. :-) Marked as deferred.
